mobile-browser-based-version/src/components/building_frames/containers/TrainingFrame.vue

```diff
@@ -125,27 +139,63 @@ export default {
   data() {
     return {
       isConnected: false,
-      // Assist with the training loop
-      trainingManager: null,
-      // Manager that returns feedbacks when training
+      startedTraining: false,
+      isTraining: false,
+      distributedTraining: false,
+      // Delivers training feedback to the user
       trainingInformant: new TrainingInformant(10, this.Task.taskID),
-      // Manager for the file uploading process
+      // Handles the file uploading process
       fileUploadManager: new FileUploadManager(this.nbrClasses, this),
-      // Take care of communication processes
-      client: getClient(
-        this.$store.getters.platform,
-        this.Task,
-        this.$store.getters.password(this.Id)
-      ), // TO DO: to modularize
     };
   },
   methods: {
+    async connectClientToServer() {
+      this.isConnected = await this.client.connect();
+      if (this.isConnected) {
+        this.$toast.success(
+          'Succesfully connected to server. Distributed training available.'
+        );
+      } else {
+        console.log('Error in connecting');
+        this.$toast.error(
+          'Failed to connect to server. Fallback to training alone.'
+        );
+      }
+      setTimeout(this.$toast.clear, 30000);
+    },
     goToTesting() {
       this.$router.push({
         path: 'testing',
       });
     },
-    async saveModelButton() {
+    async stopTraining() {
+      /**
+       * Stopping distributed training in a clean manner requires many
+       * additions and changes to make everything robust.
```

Comment:
These changes are not w.r.t telling the server that we are leaving right? (Since the server is taking care of checking for timeouts as you mention).

What currently missing for it to be more robust?

Reply:
Right, even if _ideally_ it should still include a disconnection process.

What's missing is to interrupt all ongoing processes made by a call to model.fit. This is to prevent having multiple model.fit calls running simultaneously (typically when the user resumes training), which causes an error on the side of TF.js. It can easily happen as model.fit makes API calls to the server which sometimes are performed many times (selection + aggregation status). Hence, to make it robust, one should add code to fully stop (on pressing the "stop" button) every single ongoing process after a call to model.fit.

Another option would be for the stopping process to wait on all API calls to complete (either succeeding or failing), instead of actively trying to interrupt those calls. This is counter-intuitive from a user's POV, since pressing the "stop" button would not immediately stop the training process. Furthermore, in the case of waiting for a "selection status" call to finally complete, this would make the remote server select the client, and then this very client would directly disconnect as it is trying to stop the training process. Hence, it would grief the federated training overall.

This was much easier in the local setting, as all model.fit does here is gradient updates at the end of every epoch, so the "stop" button has very little lag.

Reply:
Oh and yes, all these struggles should not occur with the **asynchronous** version of federated training.

Reply:
Yeah, waiting for the current round to finish instead of trying to stop on-going calls seems much cleaner from a dev POV.